Extend payload join optimization for cases without map columns; fix b…

mlyublena opened the pull request:
A collection of bug fixes for the payload join optimization, including the following:
+ fixed cases where not all collected join keys belong to the left-most table
+ abort optimization if the rewrite hides required columns
+ abort rewrite for operators other than Select/Project/Scan/Join. This fixes wrong result bugs when the query contains aggregates and other operators that may change the cardinality of intermediate levels
+ remove restriction for the fact table to have map or array columns. In the future we'll incorporate HBO to make a cost-based decision when to apply this optimization

rschlussel reviewed (commented):
Still reviewing the second commit.  Can you add more context to the commit message about what the problem is/why it's a problem?  

Also, can you make sure all of your commit messages follow the guidelines here: https://cbea.ms/git-commit/

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -244,9 +251,9 @@ public PlanNode visitJoin(JoinNode joinNode, RewriteContext<JoinContext> context
         {
             if (joinType == JoinNode.Type.LEFT && supportedJoinKeyTypes(joinKeys)) {
-                Set<VariableReferenceExpression> leftPayloadColumns = extractPayloadColumns(leftColumns);
+                boolean hasPayloadColumns = leftColumns.stream().anyMatch(var -> !joinKeys.contains(var));
```
that was the original intent but @kaikalur suggested we always do it as long as the flag is on.
I imagine in the future we can make this a cost-based decision based on size/types of payload columns

kaikalur replied:
> that was the original intent but @kaikalur suggested we always do it as long as the flag is on. I imagine in the future we can make this a cost-based decision based on size/types of payload columns

This is a general thing we do with all our optimizations - if the flag is on, we will do it! And yes hopefully in the next half we can have HBO help us turn it on automatically as needed.

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -1928,15 +1928,22 @@ public boolean isOptimizeNullsInJoin()
         return optimizeNullsInJoin;
     }
 
+    @Config("optimize-nulls-in-join")
```
moved feature config fixes to a separate PR

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -192,16 +186,31 @@ public PlanNode visitUnion(UnionNode unionNode, RewriteContext<JoinContext> cont
         }
 
+        @Override
+        public PlanNode visitAggregation(AggregationNode aggNode, RewriteContext<JoinContext> context)
```
generally avoid abbreviations.  aggNode -> aggregationNode or aggregation

mlyublena replied:
fixed

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -192,16 +186,31 @@ public PlanNode visitUnion(UnionNode unionNode, RewriteContext<JoinContext> cont
             ImmutableSet<VariableReferenceExpression> leftColumns = leftNode.getOutputVariables().stream().collect(toImmutableSet());
+
+            // remove join keys coming from RHS of join
+            ImmutableSet<VariableReferenceExpression> redundantJoinKeys = oldJoinKeys.stream().filter(key -> rightNode.getOutputVariables().contains(key)).collect(toImmutableSet());
```
calling these redundant makes this sound like an optimization. but as i understand it's a correctness/valid plan issue.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -238,6 +247,8 @@ public PlanNode visitJoin(JoinNode joinNode, RewriteContext<JoinContext> context
 
             if (isTopJoin && context.get().needsPayloadRejoin()) {
                 PlanNode payloadJoin = transformJoin(newJoinNode, joinContext);
+                context.get().setPayloadNode(null);
```
can you explain why the payloadNode gets set to null here?

mlyublena replied:
added a comment

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -192,16 +186,31 @@ public PlanNode visitUnion(UnionNode unionNode, RewriteContext<JoinContext> cont
+
+            // remove join keys coming from RHS of join
+            ImmutableSet<VariableReferenceExpression> redundantJoinKeys = oldJoinKeys.stream().filter(key -> rightNode.getOutputVariables().contains(key)).collect(toImmutableSet());
+            joinContext.removeKeys(redundantJoinKeys);
```
is there any concern that this will affect the isTopJoin value on future visits to this method (for nested joins)?

mlyublena replied:
I decided against supporting this case. Fixed the code to abort rewrite if the collected join keys contain an element from the RHS of the current join

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -250,13 +261,7 @@ public PlanNode visitJoin(JoinNode joinNode, RewriteContext<JoinContext> context
-            }
-
-            return false;
+            return joinType == JoinNode.Type.LEFT && supportedJoinKeyTypes(joinKeys) && leftColumns.stream().anyMatch(var -> !joinKeys.contains(var));
```
static import LEFT

rschlussel reviewed (commented):
I find the flow with all the context setting a bit confusing (even before this change), so I have a hard time being confident about the change.

It would be good to have more tests for cases that we don't expect the optimizer to kick in (e.g. left join with nested right or inner joins). Also, looks like the tests for correctness just validate the row count, but not the values or even column types.  Would be good to have tests do a regular assertQuery correctness test for all the test queries to fully validate the results.

mlyublena commented:
> I find the flow with all the context setting a bit confusing (even before this change), so I have a hard time being confident about the change.
> 
> It would be good to have more tests for cases that we don't expect the optimizer to kick in (e.g. left join with nested right or inner joins). Also, looks like the tests for correctness just validate the row count, but not the values or even column types. Would be good to have tests do a regular assertQuery correctness test for all the test queries to fully validate the results.

The logic is basically as follows:
- when visiting nodes on the way down I'm collecting join keys in the join context until I hit a supported leaf node (SelectProjectScan node)
- if all join keys are contained are contained in the leaf node, I add an Aggregate on top to compute only the join keys without the payload, and add the expression to reattach back the payload in the context
- some of the intermediate nodes may require columns from the payload table that are not join keys (and the Aggregate hid them): if that is the case I abort the optimization and try from scratch with a new context. I refactored this logic to use a utility function usedColumns that returns the necessary columns for each node type, and use that in validateUsedColumnsForUnaryNode
- when I reach back the root of the join stack (isTopJoin = true) I reattach the payload using the IsNotDistinctFrom join. I also reset the context in that case to mark that the payload part has been consumed and doesn't need to get reattached

Regarding testing, I'd prefer to create a verifier suite (like this one that I used to debug the remaining issues: https://www.internalfb.com/intern/presto/verifier/results/?suite=test_payload_0808) as it gets harder and harder to enumerate all combinations of joins and unary operators

mlyublena commented:
Note: will squash the commits after review and will update the commit description

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -481,6 +521,57 @@ private boolean supportedJoinKeyTypes(Set<VariableReferenceExpression> joinKeys)
         }
+
+        // TODO: move this as a member function of PlanNode
+        private HashSet<VariableReferenceExpression> usedColumns(PlanNode plan)
```
Return type should be Set<> rather than the specific implementation type

rschlussel replied:
i think this can be static

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -481,6 +521,57 @@ private boolean supportedJoinKeyTypes(Set<VariableReferenceExpression> joinKeys)
+                AggregationNode aggregationNode = (AggregationNode) plan;
+                List<VariableReferenceExpression> groupingCols = aggregationNode.getGroupingKeys();
+                Collection<AggregationNode.Aggregation> aggregations = aggregationNode.getAggregations().values();
+                List<VariableReferenceExpression> aggregationCols = new ArrayList<>();
```
```
List<VariableReferenceExpression aggregationCols = aggregations.stream()
       .flatMap(aggregation -> getColumnsInAggregationArguments(aggregation).stream())
       .collect(toImmutableList());
 ```

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -481,6 +521,57 @@ private boolean supportedJoinKeyTypes(Set<VariableReferenceExpression> joinKeys)
+        private List<VariableReferenceExpression> getColumnsInAggregationArguments(AggregationNode.Aggregation aggregation)
+        {
+            List<VariableReferenceExpression> result = new ArrayList<>();
+            aggregation.getArguments().forEach(argument -> result.addAll(VariablesExtractor.extractAll(argument)));
```
create an immutableList to return instead
```
return aggregation.getArguments().stream()
       .flatMap(VariablesExtractor.extractAll(argument).stream())
       .collect(toImmutableList());
```

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -455,6 +479,22 @@ private boolean validateProjectAssignments(ProjectNode projectNode)
+                return planNode;
+            }
+
+            HashSet<VariableReferenceExpression> usedColumns = usedColumns(planNode);
```
declared type should just be Set<>

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -481,6 +521,57 @@ private boolean supportedJoinKeyTypes(Set<VariableReferenceExpression> joinKeys)
+            return result;
+        }
+
+        private List<VariableReferenceExpression> getColumnsInAggregationArguments(AggregationNode.Aggregation aggregation)
```
i think this can be static

rschlussel reviewed (commented):
> some of the intermediate nodes may require columns from the payload table that are not join keys (and the Aggregate hid them): if that is the case I abort the optimization and try from scratch with a new context. I refactored this logic to use a utility function usedColumns that returns the necessary columns for each node type, and use that in validateUsedColumnsForUnaryNode

I wonder if it's possible to have the initial rewrite take into account the required columns to begin with.  It seems not ideal that the rewrite can produce a "wrong" plan that we need to undo afterwards.


> Regarding testing, I'd prefer to create a verifier suite (like this one that I used to debug the remaining issues: https://www.internalfb.com/intern/presto/verifier/results/?suite=test_payload_0808) as it gets harder and harder to enumerate all combinations of joins and unary operators

The verifier suite is great, but it's only available internally at Meta, and also doesn't run automatically with every PR as someone develops.  Of course it's impossible to be 100% comprehensive, but we do also need sufficient correctness tests in the codebase itself (of both the plan and the results).

mlyublena commented:
> > some of the intermediate nodes may require columns from the payload table that are not join keys (and the Aggregate hid them): if that is the case I abort the optimization and try from scratch with a new context. I refactored this logic to use a utility function usedColumns that returns the necessary columns for each node type, and use that in validateUsedColumnsForUnaryNode
> 
> I wonder if it's possible to have the initial rewrite take into account the required columns to begin with. It seems not ideal that the rewrite can produce a "wrong" plan that we need to undo afterwards.
> 

during top-down traversal when we explore a node we might not have encountered all the join keys yet. So we won't know if a column used in an operator is a join key from below, or is not a join key and will be hidden by the rewrite.
Consider for example the following query: Join(Filter A=5 (join(join(R,S)), T)
When we explore the filter node  we know what the top level join keys are but not what the join keys of the children are as we haven't traversed them yet. So we won't know if A is a key or not.

rschlussel commented:
> during top-down traversal when we explore a node we might not have encountered all the join keys yet. So we won't know if a column used in an operator is a join key from below, or is not a join key and will be hidden by the rewrite.
> Consider for example the following query: Join(Filter A=5 (join(join(R,S)), T)
> When we explore the filter node we know what the top level join keys are but not what the join keys of the children are as we haven't traversed them yet. So we won't know if A is a key or not.

But we'll know when we explore the child joins that A is a needed output column for a later node, so we should be able to choose not to rewrite at that point.

mlyublena commented:
> > during top-down traversal when we explore a node we might not have encountered all the join keys yet. So we won't know if a column used in an operator is a join key from below, or is not a join key and will be hidden by the rewrite.
> > Consider for example the following query: Join(Filter A=5 (join(join(R,S)), T)
> > When we explore the filter node we know what the top level join keys are but not what the join keys of the children are as we haven't traversed them yet. So we won't know if A is a key or not.
> 
> But we'll know when we explore the child joins that A is a needed output column for a later node, so we should be able to choose not to rewrite at that point.

That's a great point - I ran into this in another optimization where I wanted to know the required columns from the top node (and that one was a rule-based optimization so things were even harder). I think it's worth having built-in support for passing required properties (such as required columns) from parent operators and making it a first-class citizen in the optimizer.

mlyublena commented:
> It would be good to have more tests for cases that we don't expect the optimizer to kick in (e.g. left join with nested right or inner joins). Also, looks like the tests for correctness just validate the row count, but not the values or even column types. Would be good to have tests do a regular assertQuery correctness test for all the test queries to fully validate the results.

Added more tests to cover for the various operators where we abort the transformation. Also added assertQuery tests to check for correctness of the result

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -174,41 +178,45 @@ private Rewriter(Session session, Metadata metadata, TypeProvider types, Functio
+        public PlanNode visitPlan(PlanNode planNode, RewriteContext<JoinContext> context)
         {
-            return context.defaultRewrite(node, context.get());
+            // for binary and n-ary operators, do a default rewrite with a new context for each child to avoid lateral propagation of context information
```
recently fixed that, forgot to update/remove the comment

mlyublena replied:
fixed

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -481,6 +513,60 @@ private boolean supportedJoinKeyTypes(Set<VariableReferenceExpression> joinKeys)
         }
+
+        // TODO: move this as a member function of PlanNode
+        private static Set<VariableReferenceExpression> usedColumns(PlanNode plan)
```
I narrowed down the supported cases to only filter/project/scan + join, will clean up the implementation

mlyublena replied:
fixed

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java:
```diff
@@ -455,6 +471,22 @@ private boolean validateProjectAssignments(ProjectNode projectNode)
+            }
+
+            context.get().reset();
+            return context.defaultRewrite(planNode, new JoinContext());
```
ideally, we want to do this upfront and not have to discard plans after they have been computed - I'd prefer to do this in a separate PR as this one was meant to fix a large number of the original bugs with the optimization (and the optimization is disabled by default so no latency impact on production workloads at the moment)

pranjalssh reviewed (approved):
LGTM
